sql/src/main/java/io/crate/execution/dml/upsert/ShardUpsertRequest.java

```diff
@@ -253,16 +255,33 @@ public int hashCode() {
         @Nullable
         private Object[] insertValues;
 
+        /**
+         * The URI the item was read from.
+         * It is only needed on the sender to merge the write response into the related result object.
+         * DO NOT STREAM!
+         */
+        @Nullable
+        private BytesRef sourceUri;
```

Comment:
It's a bit of a hack that we're breaking encapsulation by making the `UpsertRequest` even more aware of where it's data was coming from and that this is not streamed.
Understanding in which cases what properties are set is already hard without these changes.

Would it be possible to have some kind of external streamer component on the sender/response-receiver in order to avoid bloating up the `ShardUpsertRequest` even more?

Maybe we could also consider having concrete requests for different cases. Maybe with some kind of shared / core properties.

I don't see this as a blocker, but I think it's something we should get back to and improve

Reply:
I was also thinking about storing the sourceUri-per-item into a different data structure which would be then used instead of the request. what do you think?